### PR TITLE
Switch to MkDocs

### PR DESCRIPTION
**Pull request by MatthewCroughan**
Draft for MkDocs transition, there are various things that need to be fixed still, since the conversion was lossy. This draft exists so people can view my progress:

# Things that still need to be fixed

## Formatting ( The way the text appears on the page )
- [x] `reference-manual/` (Excluding `reference-manual/qemu` and `reference-manual/boards`)
- [x] `getting-started/`
- [ ] `user-guide/`
- [ ] `tutorial/`
   
## Pages that need reform

These pages are the main reason for the switch. We need to program the Qemu and Boards sections by creating a template, something which Sphinx was unable to accomplish, but MkDocs is. 

- [ ] `reference-manual/qemu`
- [ ] `reference-manual/boards`

## References
- [ ] Internal references to pages
    Pandoc was unable to automate this, so I need to fix the existing references. One idea I have is to have terms automatically highlighted/cross-referenced by using [abbreviations](https://squidfunk.github.io/mkdocs-material/reference/abbreviations/). This means less work when writing anyway, as you won't need to remember to cross-reference every time `Targets` or `Factory` is written for example.
- [ ] Some image embeds
   The image embeds are included using html figures rather than the MkDocs figure directive due to the automated conversion, **this is not necessary to fix** but would be nice to fix.
- [ ] Asciinema player embedding
    The Asciinema player can be included since you can use discrete inline HTML in MkDocs, a nice benefit.
- [x] Rebasing commits on master (converting the some of the PRs the have been made since starting)

# Visual Difference

The MkDocs `material` theme is currently being used. This can be changed to the `readthedocs` theme, which matches our existing documentation if necessary, so there will be no visual difference. 

image

# Notable Changes

### Factory Definition 

Was previously hard to read, did not specify the type and was abusing .rst indentation. Now, it follows the same style that most people use for describing their schemas, in `x.y.z` format.

image

### CODE REVIEW

File: mkdocs/docs/tutorials/cloud.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Cloud
+
+aws google ibm
```
Comment (mwasilew) [inline mkdocs/docs/tutorials/cloud.md:3]: Content missing?

File: mkdocs/docs/reference-manual/boards/apalis-imx6-prepare.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
+# Preparation
```
Comment (mwasilew) [inline mkdocs/docs/reference-manual/boards/apalis-imx6-prepare.md:1]: This should be a subsection of apalis-imx6.md

**Review by mwasilew — COMMENTED** [review]
I think a change needs to be done to the 'boards' section. Each board should have either it's own page with 'details', 'preparation' and 'flashing' sections or these pages should be in the board separate directory. Right now navigation in the boards section looks broken.

**Comment by angolini** [comment]
@MatthewCroughan, is there a way I can "build" it locally on my machine?

I assumed it is only a `mkdocs serve` but I faced errors so I might be missing something

**Comment by mwasilew** [comment]
It worked for me. What errors did you get?

**Comment by MatthewCroughan** [comment]
@angolini The build instructions are in the new README.md. I'll paste it here for fun:

# Foundries.io Documentation

microPlatform documentation source for https://docs.foundries.io

# Development Environment

There are 3 ways to get started developing this documentation. Each method
provides `mkdocs` with various plugins as defined in `requirements.txt`

## Nix

Clone this repository and run `nix-shell` at the root.

1. `git clone https://github.com/foundriesio/docs/`
2. `cd docs/`
3. `nix-shell`

Then run `mkdocs serve` or use the other tools provided in this environment.

## Docker

Clone this repository and run a container on the root.

Replace `<your-git-directory>` with the path to this git repository as you have
cloned it.

1. `git clone https://github.com/foundriesio/docs/`
2. `cd docs/`
3. `docker run -v <your-git-directory>:/git matthewcroughan/mkdocs-foundries-env`

## Virtualenv

Ensure you have `python3` and `python3-virtualenv` installed.

1. `git clone https://github.com/foundriesio/docs/`
2. `cd docs/`
3. Run the following:

```bash
$ virtualenv -p /usr/bin/python3 venv
$ . ./venv/bin/activate
$ pip install -r requirements.txt
```

Then run `mkdocs serve`

**Comment by MatthewCroughan** [comment]
> I think a change needs to be done to the 'boards' section. Each board should have either it's own page with 'details', 'preparation' and 'flashing' sections or these pages should be in the board separate directory. Right now navigation in the boards section looks broken.

Yes, the board section and the qemu section are the main reason I'm advocating for the switch. It's not possible in Sphinx to code these sections up and conditionally include files, such that we have one template that each board uses. This has to wait until I've fixed the formatting in the other sections though.

The steps I'm following are:
1. Fix formatting 
2. Fix references
3. Reimplement pages that need redesigns (Like boards/qemu)

**Comment by angolini** [comment]
I'm not sure if this phase (the migration) is the best to also change the way the files are organized. I would be more worried about not loosing content now, and adjust anything later.

And I agree, the main point is to be able to use variables (we will need a bunch of them) and a better way to template some pages (such as board and qemu)

**Comment by angolini** [comment]
@MatthewCroughan what are you looking for on this review? Should I be looking for which kind of things? Missing pages? Formatting? Overall look'n'feel?

**Comment by MatthewCroughan** [comment]
> @MatthewCroughan what are you looking for on this review? Should I be looking for which kind of things? Missing pages? Formatting? Overall look'n'feel?

Formatting, though you see in https://github.com/foundriesio/docs/pull/218/commits/06524dc070990cb0b3fe154d54ee8855bb02e2dc that I have also overhauled the look-and-feel of the factory-definition. The reason for this is because there was no way to do it in the way we were via .rst anyway, so it makes some sense to just start again when the section is otherwise broken. The same will be true of  `reference-manual/boards` as well as `reference-manual/qemu`.

Right now, I'm just going through and fixing these formatting issues, it is probably not a good time for review. Though, if you want to submit PRs to this branch, you can. I will be working on the `reference-manual/` folder until a commit message indicates I am done with it, I will also comment.

That means that the `user-guide` and `tutorials` folder can be worked on without interrupting me, if you feel like working on that.

**Comment by MatthewCroughan** [comment]
As part of the rebase I just pushed, the reference manual has been reformatted, next is the tutorials/user-guide sections.

File: mkdocs/docs/tutorials/cloud.md (hunk 1 above)
Comment (MatthewCroughan) [reply mkdocs/docs/tutorials/cloud.md:3]: These are the old index files that are no longer needed which are removed in https://github.com/foundriesio/docs/pull/218/commits/44a9f3502c417b15cf4331b5cb4edddcd42ccac4

**Comment by doanac** [comment]
Docs for 18cbc94662ddc8e9f3ba2759789b6869a765b72b are browesable at: https://ci.foundries.io/projects/fio-docs/builds/937/mkdocs/artifacts/index.html

**Comment by doanac** [comment]
@MatthewCroughan - i've added CI for mkdocs. I'm seeing one issue with the CI of this. I tried to set `strict: true` to make mkdocs validate our links and it does not seem to have any affect. Not sure how you are managing the TODOs for this, but I'd add that onto the list "make sure broken links break CI"

**Comment by doanac** [comment]
I also assume you need me to update fioctl to product markdown manual pages?

**Comment by MatthewCroughan** [comment]
@doanac 

> @MatthewCroughan - i've added CI for mkdocs. I'm seeing one issue with the CI of this. I tried to set `strict: true` to make mkdocs validate our links and it does not seem to have any affect. Not sure how you are managing the TODOs for this, but I'd add that onto the list "make sure broken links break CI"

I believe `mkdocs -s` will actually work, instead of putting `strict: true` into the yml file, they are different in an undocumented way.

There are a lot of utilities we can use to check links as a post-processing step, (something ran after `mkdocs build`) rather than something that happens in MkDocs natively. This is the style of MkDocs generally speaking. It leaves post-processing and pre-processing up to the user, rather than doing it all itself in a half-baked manner:

- https://www.npmjs.com/package/markdown-link-check
- https://github.com/magicmatatjahu/milv

> I also assume you need me to update fioctl to product markdown manual pages?

Yeah, that would be great. The cobra argument parsing library that Fioctl uses can do it [like this](https://github.com/spf13/cobra/blob/eb3b6397b1b5d1b0a2cd66a9afe0520f480c0a87/doc/md_docs.md) using `GenMarkdownTree` instead of `GenReSTTreeCustom`

**Comment by MatthewCroughan** [comment]
Some thoughts on CI are that I'd like to have the ability to generate production docs locally.

So, to do that, I'm going to have every function first query whether `PRODUCTION=1` is set in the environment. If it is, then it will execute and return a proper value. If not then it will place a null value into the docs. 

Right now, the return value of functions defined in `mkdocs/main.py` are able to be referenced from within the docs like this:

#### `mkdocs/docs/something.md`

The return at the time of writing is `v0.16`

```md
The current Fioctl version is {{ fioctl_version() }}
```


#### `mkdocs/main.py`
```py

import os
# Some functions to insert in to the docs via Jinja2. For example:
# {{fioctl_version() }} will place the fioctl_version number into the markdown
# file in which it is called.

def define_env(env):
  "Functions for MkDocs Jinja2"

  @env.macro
  def fioctl_version():
      import urllib.request, json

      url = "https://api.github.com/repos/foundriesio/fioctl/releases/latest"
      response = urllib.request.urlopen(url)
      data = json.loads(response.read())

      return data['tag_name']

**Comment by doanac** [comment]
> Some thoughts on CI are that I'd like to have the ability to generate production docs locally.

Maybe take the bulk of whats in the unit-test for this and move it to a `gen-docs` shell script. i'll do a quick change with that to get the ball rolling for you.

**Comment by doanac** [comment]
> I believe mkdocs -s will actually work, instead of putting strict: true into the yml file, they are different in an undocumented way.

Neither works. I've added this to index.md:
~~~
[this is bad](http://blomberg.commmm)
~~~

and it just renders it.

**Comment by doanac** [comment]
Docs for 302ed84a90c16cb372143c1d44d5bc49af38703e are browesable at: https://ci.foundries.io/projects/fio-docs/builds/938/mkdocs/artifacts/index.html

**Comment by doanac** [comment]
Docs for 62bebff01e323f6e48dfab148497cdd880c5f15c are browesable at: https://ci.foundries.io/projects/fio-docs/builds/940/mkdocs/artifacts/index.html

**Comment by doanac** [comment]
Docs for 960faca2ce1ee8a6409b2a1361f1139eb89bc290 are browesable at: https://ci.foundries.io/projects/fio-docs/builds/941/mkdocs/artifacts/index.html

**Comment by MatthewCroughan** [comment]
I've rebased this PR/Branch on master, coming up to date with the latest changes and converting .rst -> .md

At this point, people can contribute by making PRs against this `mkdocs` branch. The `tutorials` folder is completely untouched by me. This section and its contents are what I need the most help with. I've edited the [first post in this PR](https://github.com/foundriesio/docs/pull/218#issue-640693740) to contain a checklist of what still needs to be done.

**Comment by doanac** [comment]
Docs for 9c37a25a2c562be6b08a840cdf5053bb922da248 are browesable at: https://ci.foundries.io/projects/fio-docs/builds/942/mkdocs/artifacts/index.html

**Comment by doanac** [comment]
Docs for 18a1a7ad683646e868bb7562bdc6b07d87bb5010 are browesable at: https://ci.foundries.io/projects/fio-docs/builds/950/mkdocs/artifacts/index.html

**Comment by doanac** [comment]
Docs for e78566910565cb1d32fd5ae0bd9d9afbcb8bb409 are browesable at: https://ci.foundries.io/projects/fio-docs/builds/994/mkdocs/artifacts/index.html

**Pull request closed without merging**